drivers/mil_passive_sonar/scripts/multilateration

```diff
@@ -57,50 +66,61 @@ class MultilaterationNode(object):
         except tf2_ros.TransformException, e:
             rospy.logwarn('TF Exception: {}'.format(e))
             return
+        if self.last_pos is not None and np.linalg.norm(self.last_pos - origin) < 0.2:
+            rospy.loginfo("Observation too close to previous, ignoring")
+            return
+        self.last_pos = origin
         self.vec_samples.append((origin, origin + vec))
-        if len(self.vec_samples) == self.vec_samples.maxlen:
+        if len(self.vec_samples) >= self.buffer_min:
             self.publish_position()
 
     def publish_position(self):
-        line_array = np.array([(p[0][0], p[0][1], p[1][0], p[1][1]) for p in self.vec_samples])
+        #line_array = np.array([(p[0][0], p[0][1], p[1][0], p[1][1]) for p in self.vec_samples])
+        line_array = np.array([(np.array(p[0][0:2]), np.array(p[1][0:2])) for p in self.vec_samples])
         where = self.ls_line_intersection2d(line_array)
         ps = PointStamped()
+        ps.header.frame_id = self.global_frame
         ps.header.stamp = rospy.Time.now()
         ps.point = numpy_to_point(where)
         self.position_pub.publish(ps)
 
-    def ls_line_intersection2d(self, line_array):
-        """
-        Find the intersection of lines in the least-squares sense.
-        start - Nx3 numpy array of start points
-        end - Nx3 numpy array of end points
-        https://en.wikipedia.org/wiki/Line-line_intersection#In_two_dimensions_2
-        """
-        def line_segment_norm(line_end_pts):
-            assert len(line_end_pts) == 4
-            return np.linalg.norm(line_end_pts[2:] - line_end_pts[:2])
-
-        begin_pts = line_array[:, :2]
-        diffs = line_array[:, 2:4] - begin_pts
-        norms = np.apply_along_axis(line_segment_norm, 1, line_array).reshape(diffs.shape[0], 1)
-        rot_left_90 = np.array([[0, -1], [1, 0]])
-        perp_unit_vecs = np.apply_along_axis(lambda unit_diffs: rot_left_90.dot(unit_diffs), 1, diffs / norms)
-        A_sum = np.zeros((2, 2))
-        Ap_sum = np.zeros((2, 1))
-
-        for x, y in zip(begin_pts, perp_unit_vecs):
-            begin = x.reshape(2, 1)
-            perp_vec = y.reshape(2, 1)
-            A = perp_vec.dot(perp_vec.T)
-            Ap = A.dot(begin)
-            A_sum += A
-            Ap_sum += Ap
-
-        return np.linalg.inv(A_sum).dot(Ap_sum)
+    def ls_line_intersection2d(self, lines_array):
```

Comment:
Just want to point out that what you're doing here (average of intersections of all possible pairs) yields a mathematically suboptimal estimate of the position of the source. Throwing out observations that are too close is sensible but swapping out least squares intersection for this average is not. Consider using this as a preprocessing step to filter the array of lines.